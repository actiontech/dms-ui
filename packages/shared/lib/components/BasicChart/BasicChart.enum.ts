export enum ChartTypeEnum {
  line = 'line',
  column = 'column'
}
