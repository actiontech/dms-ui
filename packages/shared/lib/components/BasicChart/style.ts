import { styled } from '@mui/material/styles';
import { Card } from 'antd5';
import { ANTD_PREFIX_STR } from '../../data/common';

export const BasicLineChartStyleWrapper = styled(Card)`
  &.${ANTD_PREFIX_STR}-card.${ANTD_PREFIX_STR}-card-bordered {
    border-radius: 8px;
    border: 1px solid ${({ theme }) => theme.sharedTheme.basic.colorGrayLine};
    padding: 16px 20px;

    .${ANTD_PREFIX_STR}-card-body {
      padding: 0;
      border-radius: none;
    }
  }
`;
