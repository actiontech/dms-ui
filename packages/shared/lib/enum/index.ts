export enum PackageNameEnum {
  SQLE = 'sqle',
  PROVISION = 'provision'
}

export enum SupportLanguage {
  zhCN = 'zh-CN',
  enUS = 'en-US'
}

export enum ResponseCode {
  SUCCESS = 0,

  // sql analyze
  NotSupportDML = 8001,

  // version management
  BatchTaskNotFullyCompleted = 8005
}

export enum SupportTheme {
  DARK = 'dark',
  LIGHT = 'light'
}

export enum SystemRole {
  admin = 'admin',
  certainProjectManager = 'certainProjectManager',
  globalViewing = 'globalViewing'
}

export type UserRolesType = {
  [key in SystemRole]: boolean;
};

export enum ModalSize {
  big = 1000,
  mid = 800
}

export enum CharCode {
  a = 97,
  z = 122,
  A = 65,
  Z = 90
}

export enum StorageKey {
  Language = 'LANGUAGE',
  Theme = 'THEME',
  Token = 'TOKEN',
  DMS_Project_Catch = 'DMS_Project_Catch',
  USER_UID = 'DMS_USER_UID',
  SHOW_COMPANY_NOTICE = 'SHOW_COMPANY_NOTICE',
  // diagnosis user id
  DIAGNOSIS_USER_ID = 'DIAGNOSIS_USER_ID'
}

/**
 * 后端暂时无法在swagger中暴露OpPermissionType的映射，但权限对应的uid一般不会变化，因此前端先保存一份。
 */
export enum OpPermissionTypeUid {
  'create_project' = '700001', // 创建项目；创建项目的用户自动拥有该项目管理权限 700001
  'project_admin' = '700002', // 项目管理；拥有该权限的用户可以管理项目下的所有资源 700002
  'create_workflow' = '700003', // 创建/编辑工单；拥有该权限的用户可以创建/编辑工单 700003
  'audit_workflow' = '700004', // 审核/驳回工单；拥有该权限的用户可以审核/驳回工单 700004
  'auth_db_service_data' = '700005', // 授权数据源数据权限；拥有该权限的用户可以授权数据源数据权限 700005
  'execute_workflow' = '700006', // 上线工单；拥有该权限的用户可以上线工单 700006
  'view_others_workflow' = '700007', // 查看其他工单权限 700007
  'view_other_audit_plan' = '700008', // 查看其他扫描任务权限 700008
  'save_audit_plan' = '700009', // 创建扫描任务权限；拥有该权限的用户可以创建/更新扫描任务 700009
  'sql_query' = '700010', //SQL查询；SQL查询权限 700010
  'global_viewing' = '700016', // 全局浏览
  'global_management' = '700017' // 全局管理
}

export enum CompanyNoticeDisplayStatusEnum {
  Displayed = '1',
  NotDisplayed = '0'
}

export enum MIMETypeEnum {
  'Application_Json' = 'application/json'
}

export type ResponseBlobJsonType = {
  code: number;
  message: string;
};

export enum KeyboardType {
  Enter = 'Enter'
}
