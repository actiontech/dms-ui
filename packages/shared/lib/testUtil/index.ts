export * from './superRender';
export * from './mockHook';
export * from './mockApi';
export * from './customQuery';
export * from './mockRedux';
export * from './common';
export * from './LocationDisplay';
