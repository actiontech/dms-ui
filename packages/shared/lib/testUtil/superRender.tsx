import { render, renderHook as renderHookReact } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MutableSnapshot, RecoilRoot } from 'recoil';
import { MemoryRouter, MemoryRouterProps } from 'react-router-dom';
import { ConfigProvider, theme as antdTheme } from 'antd';
import zhCN from 'antd/locale/zh_CN';
import { storeFactory } from './mockRedux';
import { ThemeProvider } from '@mui/system';
import { StyledEngineProvider } from '@mui/material/styles';
import lightTheme from '../theme/light';

export type RenderParams = Parameters<typeof render>;

export type RecoilRootPropsCustom = {
  initializeState?: (mutableSnapshot: MutableSnapshot) => void;
  override?: boolean;
};

export interface WrapperOptions {
  routerProps?: MemoryRouterProps;
  initStore?: any;
  recoilRootProps?: RecoilRootPropsCustom;
  theme?: any;
}

export const createTestWrapper = (options?: WrapperOptions) => {
  const {
    routerProps,
    initStore,
    recoilRootProps,
    theme = lightTheme
  } = options || {};

  return ({ children }: { children: React.ReactNode }) => {
    const themeComponent = (
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    );

    return (
      <ConfigProvider
        locale={zhCN}
        theme={{ algorithm: antdTheme.defaultAlgorithm, hashed: false }}
      >
        <RecoilRoot {...recoilRootProps}>
          <Provider store={storeFactory(initStore)}>
            <MemoryRouter {...routerProps}>
              <StyledEngineProvider injectFirst>
                {themeComponent}
              </StyledEngineProvider>
            </MemoryRouter>
          </Provider>
        </RecoilRoot>
      </ConfigProvider>
    );
  };
};

export const superRender = (
  ...[ui, option, options]: [...RenderParams, WrapperOptions?]
) => {
  const wrapper = createTestWrapper(options);

  return render(ui, {
    wrapper,
    ...option
  });
};

export const superRenderHook = <TProps, TResult>(
  hook: (props: TProps) => TResult,
  renderHookOptions?: Omit<
    Parameters<typeof renderHookReact<TProps, TResult>>[1],
    'wrapper'
  >,
  wrapperOptions?: WrapperOptions
) => {
  const wrapper = createTestWrapper(wrapperOptions);

  return renderHookReact(hook, {
    ...renderHookOptions,
    wrapper
  });
};
