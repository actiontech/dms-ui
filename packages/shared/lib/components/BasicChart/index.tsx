import React, { useMemo } from 'react';
import { Column, ColumnConfig, Line, LineConfig } from '@ant-design/plots';
import { BasicLineChartStyleWrapper } from './style';
import BasicEmpty from '../BasicEmpty';

export enum ChartTypeEnum {
  line = 'line',
  column = 'column'
}

export interface BasicChartProps {
  type: ChartTypeEnum;
  loading?: boolean;
  errorInfo?: string | React.ReactNode;
  errorTitle?: string | React.ReactNode;
  config?: LineConfig | ColumnConfig;
  language?: string;
  theme?: string | object;
}

const BasicChart: React.FC<BasicChartProps> = (props) => {
  const { type, loading, errorInfo, errorTitle, config, language, theme } =
    props;

  const chart = useMemo(() => {
    return type === ChartTypeEnum.line ? (
      <Line {...(config as LineConfig)} theme={theme} locale={language} />
    ) : (
      <Column {...(config as ColumnConfig)} theme={theme} locale={language} />
    );
  }, [type, config, theme, language]);

  return (
    <BasicLineChartStyleWrapper>
      {loading || config?.data.length === 0 || !!errorInfo ? (
        <BasicEmpty
          loading={loading}
          dataLength={config?.data.length}
          errorInfo={errorInfo}
          errorTitle={errorTitle}
        />
      ) : (
        chart
      )}
    </BasicLineChartStyleWrapper>
  );
};

export default BasicChart;
