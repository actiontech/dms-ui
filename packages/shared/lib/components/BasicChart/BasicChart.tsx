import React, { useMemo } from 'react';
import { Column, ColumnConfig, Line, LineConfig } from '@ant-design/plots';
import { BasicLineChartStyleWrapper } from './style';
import BasicEmpty from '../BasicEmpty/BasicEmpty';
import { BasicChartProps } from './BasicChart.types';
import { ChartTypeEnum } from './BasicChart.enum';

const BasicChart: React.FC<BasicChartProps> = (props) => {
  const { type, loading, errorInfo, errorTitle, config, language, theme } =
    props;

  const chart = useMemo(() => {
    return type === ChartTypeEnum.line ? (
      <Line {...(config as LineConfig)} theme={theme} locale={language} />
    ) : (
      <Column {...(config as ColumnConfig)} theme={theme} locale={language} />
    );
  }, [type, config, theme, language]);

  return (
    <BasicLineChartStyleWrapper>
      {loading || config?.data.length === 0 || !!errorInfo ? (
        <BasicEmpty
          loading={loading}
          dataLength={config?.data.length}
          errorInfo={errorInfo}
          errorTitle={errorTitle}
        />
      ) : (
        chart
      )}
    </BasicLineChartStyleWrapper>
  );
};

export default BasicChart;
