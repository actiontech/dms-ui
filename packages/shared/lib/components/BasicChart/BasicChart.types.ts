import { LineConfig, ColumnConfig } from '@ant-design/plots';
import { ChartTypeEnum } from './BasicChart.enum';

export interface BasicChartProps {
  type: ChartTypeEnum;
  loading?: boolean;
  errorInfo?: string | React.ReactNode;
  errorTitle?: string | React.ReactNode;
  config?: LineConfig | ColumnConfig;
  language?: string;
  theme?: string | object;
}
