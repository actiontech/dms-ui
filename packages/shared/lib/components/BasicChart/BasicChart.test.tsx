import { superRender } from '../../testUtil/superRender';
import BasicChart from './BasicChart';
import { ChartTypeEnum } from './BasicChart.enum';
import { BasicChartProps } from './BasicChart.types';
import { BasicEmpty } from '@actiontech/dms-kit';
import { screen } from '@testing-library/react';

jest.mock('@actiontech/dms-kit', () => ({
  ...jest.requireActual('@actiontech/dms-kit'),
  BasicEmpty: jest.fn().mockImplementation(() => null)
}));

describe('lib/BasicChart', () => {
  const customRender = (params: BasicChartProps) => {
    return superRender(<BasicChart {...params} />);
  };

  beforeEach(() => {
    (BasicEmpty as jest.Mock).mockClear();
  });

  it('should match snapshot', () => {
    const { container } = customRender({
      type: ChartTypeEnum.line,
      config: {
        data: [{ x: 1 }]
      }
    });
    expect(container).toMatchSnapshot();
  });

  it('should render loading state correctly', () => {
    customRender({
      type: ChartTypeEnum.line,
      loading: true
    });
    expect(BasicEmpty).toHaveBeenCalledWith(
      expect.objectContaining({
        loading: true
      }),
      expect.any(Object)
    );
  });

  it('should render empty state when data is empty', () => {
    customRender({
      type: ChartTypeEnum.line,
      config: {
        data: []
      }
    });
    expect(BasicEmpty).toHaveBeenCalledWith(
      expect.objectContaining({
        dataLength: 0
      }),
      expect.any(Object)
    );
  });

  it('should render error state correctly', () => {
    const errorTitle = '这是错误信息的标题';
    const errorInfo = '这是一个错误信息';
    customRender({
      type: ChartTypeEnum.line,
      loading: false,
      errorTitle,
      errorInfo
    });
    expect(BasicEmpty).toHaveBeenCalledWith(
      expect.objectContaining({
        errorTitle,
        errorInfo
      }),
      expect.any(Object)
    );
  });

  it('should render line chart correctly', () => {
    customRender({
      type: ChartTypeEnum.line,
      theme: 'dark',
      config: {
        data: [{ x: 1 }]
      }
    });
    const lineChart = screen.getByTestId('mock-antd-plots');
    expect(lineChart).toBeInTheDocument();
    expect(lineChart.getAttribute('data-custom-params')).toEqual(
      JSON.stringify({
        data: [{ x: 1 }],
        theme: 'dark'
      })
    );
  });

  it('should render column chart correctly', () => {
    customRender({
      type: ChartTypeEnum.column,
      theme: 'dark',
      language: 'zh-cn',
      config: {
        data: [{ x: 1 }]
      }
    });
    const columnChart = screen.getByTestId('mock-antd-plots');
    expect(columnChart).toBeInTheDocument();
    expect(columnChart.getAttribute('data-custom-params')).toEqual(
      JSON.stringify({
        data: [{ x: 1 }],
        theme: 'dark',
        locale: 'zh-cn'
      })
    );
  });
});
