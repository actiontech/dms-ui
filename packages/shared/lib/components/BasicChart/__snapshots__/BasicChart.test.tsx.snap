// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/BasicChart should match snapshot 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered css-qsi9tz"
  >
    <div
      class="ant-card-body"
    >
      <div
        data-custom-params="{"data":[{"x":1}]}"
        data-testid="mock-antd-plots"
      />
    </div>
  </div>
</div>
`;
